Add OpenMote-CC2538 platform and examples.

peretuset opened the pull request:
After a long time, the OpenMote-CC2538 platform is finally ready to be merged into the Contiki's main repository. I have tested all the platform-related code and also the examples and they all work allright.
Please review the PR and make suggestions to improve it if required. Thanks!

simonduq commented:
Thanks the lot for bringing this back to life!!
Minor things that need to be addressed:
* fix doxygen warnings (the reason why Travis failed)
* comply with https://github.com/contiki-os/contiki/wiki/Code-Contributions#new-platforms , in particular
 * add content to the readme
 * name at least one long-term maintainer

simonduq commented:
Oh I forgot one:
* add compile non-regression tests (very easy, just add a few lines to https://github.com/contiki-os/contiki/blob/master/regression-tests/18-compile-arm-ports/Makefile)

peretuset commented:
After a long night banging my head against the wall with Doxygen syntax and some other minor issues, it looks like it is now passing all the CI tests! I have also added the non-regression tests and added content to the readme file. Any further suggestions are welcome!

simonduq commented:
Looks great, fulfills the new platform requirements, and I know this has already been used by many. Gets my :+1:

alignan commented on platform/openmote-cc2538/Makefile.openmote-cc2538:
```diff
@@ -0,0 +1,50 @@
+           core/net/llsec core/net/llsec/noncoresec
+
+PYTHON = python
+BSL_FLAGS += -e -w --bootloader-invert-lines -b 115200 -v
```
It is not mandatory, but the new `cc2538-bsl` allows higher baudrates than 115200, for other platforms the `-b` argument was dropped

peretuset replied:
Indeed, thanks!

alignan commented on platform/openmote-cc2538/contiki-conf.h:
```diff
@@ -0,0 +1,573 @@
+ *
+ * @{
+ */
+#define COFFEE_CONF_SIZE            (4 * COFFEE_SECTOR_SIZE)
```
it was suggested by @bthebaudeau to change this to:

````
#ifndef COFFEE_CONF_SIZE
#define COFFEE_CONF_SIZE            (4 * COFFEE_SECTOR_SIZE)
#endif
````

peretuset replied:
Noted, thanks!

alignan commented on platform/openmote-cc2538/dev/adxl346.c:
```diff
@@ -0,0 +1,234 @@
+  return z;
+}
+/*---------------------------------------------------------------------------*/
+/** @} */
```
No newline at the end of the file

peretuset replied:
Done!

alignan commented on platform/openmote-cc2538/dev/max44009.c:
```diff
@@ -0,0 +1,190 @@
+  return result;
+}
+/*---------------------------------------------------------------------------*/
+/** @} */
```
No newline at the end of the file

peretuset replied:
Done!

alignan commented on platform/openmote-cc2538/dev/max44009.h:
```diff
@@ -0,0 +1,80 @@
+/**
+ * @}
+ * @}
+ */
```
Missing newline

peretuset replied:
Done!

alignan commented on platform/openmote-cc2538/dev/sht21.c:
```diff
@@ -0,0 +1,197 @@
+  return result;
+}
+/*---------------------------------------------------------------------------*/
+/** @} */
```
Newline missing

peretuset replied:
Done!

alignan commented on platform/openmote-cc2538/dev/sht21.h:
```diff
@@ -0,0 +1,90 @@
+/**
+ * @}
+ * @}
+ */
```
Newline missing

peretuset replied:
Done!

alignan commented:
It would be better and try to comply with Contiki's sensor API (`configure`, `value`), even if it needs a bit of rework (in my opinion), then have extended functions for features that do not fit the API.

g-oikonomou commented:
Good to see activity here again.

* There are a number of code style inconsistencies in various files, please run this through uncrustify.
* The doxygen documentation is terribly broken. There are 3 `\defgroup openmote-cc2538`. There should be only one and it should be added to group `platform`. I can see many `\addtogroup openmote`, but I cannot see the respective `defgroup` anywhere. The outcome is that the platforms documentation modules are scattered all across the html docs. This should be tidied up.
* As Antonio says, unless there is a specific reason to force speed 115200 on the BSL, allowing the default of 1000000 is recommended for faster programming.

g-oikonomou commented:
Furthermore, the convention for include guards is `THIS_H_`, so some of those need changed in this pull (don't forget the closing comment). We renamed across the board those that started with `__` a while ago and we should try to not introduce new ones.

drandreas commented on platform/openmote-cc2538/dev/antenna.c:
```diff
@@ -0,0 +1,95 @@
+antenna_external(void)
+{
+  gpio_reset(BSP_RADIO_BASE, BSP_RADIO_INT);
+  gpio_set(BSP_RADIO_BASE, BSP_RADIO_EXT);
```
In my humble opinion ```GPIO_WRITE_PIN``` should be used here.

peretuset replied:
Done, thanks!

peretuset commented:
Just got through all of your comments and the checks are still passing. Can you double check that the issues are resolved now? Thanks!

g-oikonomou commented on platform/openmote-cc2538/dev/max44009.h:
```diff
@@ -0,0 +1,80 @@
+ * \addtogroup openmote-sensors
+ * @{
+ *
+ * \defgroup openmote-max4009-sensor MAX4009 light sensor
```
Due to a typo (should be 44009 - one '4' missing), this ends up in the wrong place in the doc module hierarchy

peretuset replied:
Fixed the typo, should be ready to merge once the test are passed.
Thanks for taking care of the review!

g-oikonomou commented:
Pere, thanks. Other than the typo that I flagged inline, all my comments have been addressed and I'd be more than happy to merge this one (at long last!) :+1:

g-oikonomou commented:
:+1:

alignan commented:
I just have a comment about making the sensors compatible with current Contiki sensors implementation, but no strong objection about it (it can be done in a new PR, although this normally doesn't happen)... @g-oikonomou mind-tricked me to port all of our sensors a while ago, so perhaps he can comment :smile_cat:

g-oikonomou commented:
Even though the sensors API is a little PITA, I always choose to use it and encourage others to do so. When I read the comments I somehow thought this has been done already. Would be nice, but I'm not hugely bothered - support for this platform is long overdue

kkrentz commented:
:+1: Works, except for some issue with the 6LoWPAN adaption layer. A small change did the trick: https://github.com/kkrentz/contiki/commit/eec004d796f47d9dfc28a39ee95ae20b9aeab1bb

alignan commented:
@kkrentz in any case, it should be:
````
/* Define our IPv6 prefixes/contexts here */
#define SICSLOWPAN_CONF_MAX_ADDR_CONTEXTS    1
#ifndef SICSLOWPAN_CONF_ADDR_CONTEXT_0
#define SICSLOWPAN_CONF_ADDR_CONTEXT_0 { \
  addr_contexts[0].prefix[0] = UIP_DS6_DEFAULT_PREFIX_0; \
  addr_contexts[0].prefix[1] = UIP_DS6_DEFAULT_PREFIX_1; \
}
#endif
````
Am I wrong? or is there any other related problem for other platforms as well?

kkrentz commented:
This also works - so apparently the problem I experienced was a wrong mapping between context identifier and network prefix

g-oikonomou commented:
@peretuset This pull needs updated to reflect recent changes to 6LoWPAN contexts, as per the discussion above. The preferred approach is the one Antonio mentions, also see #1297

Also, I need some comment on whether you have any plans of updating sensor drivers to use the standard API.

peretuset commented:
Done!

g-oikonomou commented:
I'm merging this. It is still advisable to update sensor drivers to use the Contiki sensors API. But then again, this hardware and respective support have been around for too long already!

peretuset commented:
Super, thanks!
By the way, I have finished moving the sensors drivers to use the Contiki sensors API, so you can expect a new PR soon enough... :)
